### ISSUE TITLE
Support for new Cloud Functions Storage trigger events

### ISSUE THREAD

**Issue by ssaqua** [terraform-providers/terraform-provider-google#1179]
The current `providers/cloud.storage/eventTypes/object.change` trigger event type is marked as legacy with new finer-grained storage triggers available:

- `google.storage.object.finalize`
- `google.storage.object.delete`
- `google.storage.object.archive`
- `google.storage.object.metadataUpdate`

https://cloud.google.com/functions/docs/calling/storage#legacy_cloud_storage_triggers

After February 28, the legacy event type will only receive the object finalize event. It would be great to get support for these new event types although I suspect it would need a new option on the cloudfunctions resource.

Snippets from GCP email I received highlighting these changes:

> New more powerful storage triggers
We've heard your feedback and have introduced finer grained storage triggers that allow developers to invoke functions on only a subset of Cloud Storage change events as opposed to responding to all changes on a bucket. For example, you can now write a Cloud Function that only executes when a file is created or one that only executes when a file is deleted. See our documentation to learn more about how to use these new triggers. 

> Google Cloud Storage triggers
Google Cloud storage triggers will use the new finer grained event types mentioned above. As a result, the format of the events will change, and the behavior of the 'gcloud beta functions deploy --trigger-bucket' command will be changing shortly to use the new triggers. After February 28, 2018, deploying a new function with this command will no longer cause it to receive Cloud Storage object change notifications. Instead, it will cause your function to only trigger on OBJECT_FINALIZE events--when a new file is created.

> What happens to my running functions?
Your running functions will not be affected by either of these changes at this time. They will continue to receive events as usual and with the current schema. These older, legacy event types, however, are no longer being actively supported and will remain in beta indefinitely.

> What if I want to update one of my functions that is using a legacy event type?
After February 28, 2018, attempting to update a function that is using a legacy event type via the --trigger-bucket or --trigger-topic flag in gcloud will not succeed. If you want to update one of your existing functions, we have preserved the legacy Cloud Storage and legacy Pub/Sub events. These will allow you to continue to maintain your code even after the gcloud SDK behavior is changed to take advantage of the new Cloud Storage and Cloud Pub/Sub events.

### Terraform Version

`v0.11.3`

### Affected Resource(s)

- google_cloudfunctions_function

**Comment by sudomilk** [comment]
This would be a breaking change in syntax compatibility, but perhaps something like this would be the proper way of doing this?

```HCL
resource "google_storage_bucket" "example_trigger_bucket" {
    name = "test-trigger-bucket"
}

resource "google_storage_bucket" "bucket" {
  name = "test-bucket"
}

resource "google_storage_bucket_object" "archive" {
  name   = "index.zip"
  bucket = "${google_storage_bucket.bucket.name}"
  source = "./path/to/zip/file/which/contains/code"
}

resource "google_cloudfunctions_function" "function" {
  name                  = "function-test"
  description           = "My function"
  available_memory_mb   = 128
  source_archive_bucket = "${google_storage_bucket.bucket.name}"
  source_archive_object = "${google_storage_bucket_object.archive.name}"
  // event is optional 
  trigger_bucket        = {
    name  = "${google_storage_bucket.example_trigger_bucket.name}"
    event = "finalize"
  }
  timeout               = 60
  entry_point           = "helloGET"
  labels {
    my-label = "my-label-value"
  }
}
```

**Comment by naseemkullah** [comment]
bump

**Comment by paddycarver** [comment]
Is this not resolved by #2158, released in 1.19.0?

**Comment by rileykarson** [comment]
Fixed by https://github.com/GoogleCloudPlatform/magic-modules/pull/692
